scripts/revng-check-conventions

```diff
@@ -100,134 +109,244 @@ while [[ $# > 0 ]]; do
+function cat_header() {
+    local input
+    input="$(cat -)"
+
+    if echo "$input" | grep -vqz '^\s*$'; then
+        echo "$1"
+        echo "$input"
+    fi
+}
+
+
 CLANG_FORMAT_STYLE_FILE="$SCRIPT_PATH/clang-format-style-file"
-if ! test -e "$CLANG_FORMAT_STYLE_FILE"; then
+if [[ ! -e "$CLANG_FORMAT_STYLE_FILE" ]]; then
     CLANG_FORMAT_STYLE_FILE="$SCRIPT_PATH/../../share/revng/clang-format-style-file"
-    if ! test -e "$CLANG_FORMAT_STYLE_FILE"; then
+    if [[ ! -e "$CLANG_FORMAT_STYLE_FILE" ]]; then
         echo "Can't find clang-format-style-file" > /dev/stderr
         exit 1
     fi
 fi
 
 CLANG_FORMAT_STYLE=$(cat "$CLANG_FORMAT_STYLE_FILE")
 
-if test "$USE_CLANG_FORMAT_FILE" -gt 0; then
-    CLANG_FORMAT_STYLE=file
+if [[ "$USE_CLANG_FORMAT_FILE" -ne 0 ]]; then
+    CLANG_FORMAT_STYLE="file"
 fi
 
 # If the user passed the --print-clang-format-config, dump the config and exit
 # with success.
-if test "$PRINT_CLANG_FORMAT_CONFIG" -gt 0; then
+if [[ "$PRINT_CLANG_FORMAT_CONFIG" -ne 0 ]]; then
     clang-format --dry-run -style="$CLANG_FORMAT_STYLE" --dump-config
     exit 0
 fi
 
 
 if [[ $# -eq 0 ]]; then
-    FILES="$(git ls-files | grep -E '\.(c|cc|cpp|h|hpp)$')"
+    readarray -t FILES < <($GIT_LS_FILES | grep -E '\.(c|cc|cpp|h|hpp)$')
 else
-    FILES="$@"
+    FILES=("$@")
 fi
 
-GREP="git grep -n --color=always"
-
 # Run clang-format on FILES
 function run_clang_format() {
-    if test "$FORCE_FORMAT" -gt 0; then
-        clang-format -style="$CLANG_FORMAT_STYLE" -i $FILES || true
+    if [[ ${#FILES[@]} -eq 0 ]]; then return; fi
+    if [[ "$FORCE_FORMAT" -ne 0 ]]; then
+        clang-format -style="$CLANG_FORMAT_STYLE" -i "${FILES[@]}" || true
     else
-        clang-format --dry-run -style="$CLANG_FORMAT_STYLE" -i $FILES
+        clang-format --dry-run -style="$CLANG_FORMAT_STYLE" -i "${FILES[@]}"
     fi
 }
 
 # Run black
 function run_black() {
-    readarray -t PYTHON_FILES < <(( git grep -l '^#!.*python' && git ls-files | grep '\.py$' ) | sort -u )
+    readarray -t PYTHON_FILES < <($python_files)
     if [[ ${#PYTHON_FILES[@]} -eq 0 ]]; then return; fi
-    if test "$FORCE_FORMAT" -gt 0; then
+    if [[ "$FORCE_FORMAT" -ne 0 ]]; then
         black -q -l 100 "${PYTHON_FILES[@]}"
     else
-        black --check -l 100 "${PYTHON_FILES[@]}" | grep 'Oh no'
+        black --check -l 100 "${PYTHON_FILES[@]}"
     fi
 }
 
 # Run cmake-format
 function run_cmake_format() {
-    if test "$FORCE_FORMAT" -gt 0; then
-        cmake-format -i $(git ls-files | grep -i cmake) -l error
+    readarray -t CMAKE_FILES < <($GIT_LS_FILES | grep -i cmake)
+    if [[ ${#CMAKE_FILES[@]} -eq 0 ]]; then return; fi
+    if [[ "$FORCE_FORMAT" -ne 0 ]]; then
+        cmake-format -i "${CMAKE_FILES[@]}" -l error
     else
-        cmake-format --check $(git ls-files | grep -i cmake) -l error
+        cmake-format --check "${CMAKE_FILES[@]}" -l error
+    fi
+}
+
+# Run shellcheck + check for set -euo pipefail
+function run_bash_check() {
```

Comment:
Nice! But I'm really reluctant to depend on Haskell...
How do you install this on Ubuntu? On Gentoo it's quite frustrating.

Reply:
(Assuming > `18.04`)
`apt install shellcheck`
or (generally) unpack the binary archive from the [project's github release](https://github.com/koalaman/shellcheck/releases)